thread 'main' panicked at '`rustc` not found in /scratch/rustc/build/sparcv9/build/bootstrap/debug/deps, run `cargo build --bins` before `cargo run`', lib.rs:415:13

Rust 1.70 build on Solaris sometimes files on some machines:

```
Building tool rustfmt (stage2 -> stage3)
   Compiling cc v1.0.77
   Compiling camino v1.0.9
   Compiling rustfmt-nightly v1.5.2 (/scratch/psumbera/userland-test/components/rust/rustc/rustc-1.70.0-src/src/tools/rustfmt)
   Compiling yansi-term v0.1.2
   Compiling getopts v0.2.21
   Compiling unicode_categories v0.1.1
   Compiling unicode-segmentation v1.10.0
   Compiling diff v0.1.13
   Compiling bytecount v0.6.2
   Compiling dirs-sys-next v0.1.2
   Compiling dirs-sys v0.3.6
   Compiling rustfmt-config_proc_macro v0.3.0 (/scratch/psumbera/userland-test/components/rust/rustc/rustc-1.70.0-src/src/tools/rustfmt/config_proc_macro)
   Compiling derive-new v0.5.8
   Compiling env_logger v0.9.0
   Compiling cargo-platform v0.1.2
   Compiling toml v0.5.7
   Compiling dirs-next v2.0.0
   Compiling dirs v4.0.0
   Compiling annotate-snippets v0.9.1
   Compiling term v0.7.0
   Compiling cargo_metadata v0.14.0
   Compiling libz-sys v1.1.3
   Compiling rustc-workspace-hack v1.0.0 (/scratch/psumbera/userland-test/components/rust/rustc/rustc-1.70.0-src/src/tools/rustc-workspace-hack)
    Finished release [optimized] target(s) in 2m 41s
Build completed successfully in 2:11:14
Building bootstrap
    Finished dev [unoptimized] target(s) in 1.38s
warning: file locking not supported for target, not locking build directory
thread 'main' panicked at '`rustc` not found in /scratch/psumbera/userland-test/components/rust/rustc/build/sparcv9/build/bootstrap/debug/deps, run `cargo build --bins` before `cargo run`', lib.rs:415:13
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
Build completed unsuccessfully in 0:00:02
make[1]: *** [Makefile:13: all] Error 1
make[1]: Leaving directory '/scratch/psumbera/userland-test/components/rust/rustc/build/sparcv9'
'/usr/gnu/bin/make' failed with error code 2!
Running a cleanup program to restore the state.
gmake[1]: Entering directory '/scratch/psumbera/userland-test/components/rust/rustc/build/sparcv9'
Building bootstrap
    Finished dev [unoptimized] target(s) in 1.25s
warning: file locking not supported for target, not locking build directory
thread 'main' panicked at '`rustc` not found in /scratch/psumbera/userland-test/components/rust/rustc/build/sparcv9/build/bootstrap/debug/deps, run `cargo build --bins` before `cargo run`', lib.rs:415:13
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
Build completed unsuccessfully in 0:00:01
gmake[1]: *** [Makefile:25: clean] Error 1
gmake[1]: Leaving directory '/scratch/psumbera/userland-test/components/rust/rustc/build/sparcv9'
Failed to run cleanup program, terminating prematurely...
```

I wonder whether it can do something with: `warning: file locking not supported for target, not locking build directory`?!

--- a/src/bootstrap/src/lib.rs
+++ b/src/bootstrap/src/lib.rs
@@ -370,11 +370,16 @@ impl Build {
             .expect("failed to read src/version");
         let version = version.trim();
 
-        let bootstrap_out = std::env::current_exe()
+        let mut bootstrap_out = std::env::current_exe()
             .expect("could not determine path to running process")
             .parent()
             .unwrap()
             .to_path_buf();
+        // Since bootstrap is hardlink to deps/bootstrap-*, Solaris can sometimes give
+        // path with deps/ which is bad and needs to be avoided.
+        if bootstrap_out.ends_with("deps") {
+            bootstrap_out.pop();
+        }
         if !bootstrap_out.join(exe("rustc", config.build)).exists() && !cfg!(test) {
             // this restriction can be lifted whenever https://github.com/rust-lang/rfcs/pull/3028 is implemented
             panic!(

